core/src/test/java/org/hibernate/ogm/test/hint/CustomHintSupportedProvider.java

```diff
@@ -0,0 +1,179 @@
+import org.hibernate.ogm.dialect.query.spi.QueryableGridDialect;
+import org.hibernate.ogm.dialect.spi.AssociationContext;
+import org.hibernate.ogm.dialect.spi.AssociationTypeContext;
+import org.hibernate.ogm.dialect.spi.BaseGridDialect;
+import org.hibernate.ogm.dialect.spi.GridDialect;
+import org.hibernate.ogm.dialect.spi.ModelConsumer;
+import org.hibernate.ogm.dialect.spi.NextValueRequest;
+import org.hibernate.ogm.dialect.spi.OperationContext;
+import org.hibernate.ogm.dialect.spi.TupleContext;
+import org.hibernate.ogm.dialect.spi.TupleTypeContext;
+import org.hibernate.ogm.entityentry.impl.TuplePointer;
+import org.hibernate.ogm.model.key.spi.AssociationKey;
+import org.hibernate.ogm.model.key.spi.AssociationKeyMetadata;
+import org.hibernate.ogm.model.key.spi.EntityKey;
+import org.hibernate.ogm.model.key.spi.EntityKeyMetadata;
+import org.hibernate.ogm.model.spi.Association;
+import org.hibernate.ogm.model.spi.Tuple;
+import org.hibernate.ogm.query.spi.QueryParserService;
+import org.hibernate.ogm.query.spi.QueryParsingResult;
+import org.hibernate.ogm.util.impl.CollectionHelper;
+import org.hibernate.query.internal.ParameterMetadataImpl;
+
+/**
+ * @author Sergey Chernolyas &amp;sergey_chernolyas@gmail.com&amp;
+ */
+public class CustomHintSupportedProvider extends BaseDatastoreProvider {
+
+	@Override
+	public Class<? extends GridDialect> getDefaultDialect() {
+		return CustomHintSupportedProvider.Dialect.class;
+	}
+
+	@Override
+	public Class<? extends QueryParserService> getDefaultQueryParserServiceType() {
+		return CustomHintSupportedParser.class;
+	}
+
+	public static class CustomHintSupportedParser implements QueryParserService {
+
+		@Override
+		public boolean supportsParameters() {
+			return false;
+		}
+
+		@Override
+		public QueryParsingResult parseQuery(SessionFactoryImplementor sessionFactory, String queryString, Map<String, Object> namedParameters) {
+			return new CustomHintSupportedQueryParsingResult();
+		}
+
+		@Override
+		public QueryParsingResult parseQuery(SessionFactoryImplementor sessionFactory, String queryString) {
+			return new CustomHintSupportedQueryParsingResult();
+		}
+	}
+
+	public static class CustomHintSupportedQueryParsingResult implements QueryParsingResult {
+
+		@Override
+		public Object getQueryObject() {
+			return "";
+		}
+
+		@Override
+		public List<String> getColumnNames() {
+			return Collections.emptyList();
+		}
+	}
+
+	public static class Dialect extends BaseGridDialect implements QueryableGridDialect<Serializable> {
+
+		public static final String DIALECT_SPECIFIED_HINT = Dialect.class.getName() + ".hint";
+
+		public Dialect(CustomHintSupportedProvider provider) {
+		}
+
+		@Override
+		public ClosableIterator<Tuple> executeBackendQuery(BackendQuery<Serializable> query, QueryParameters queryParameters, TupleContext tupleContext) {
+			List<String> queryHints = queryParameters.getQueryHints();
+			if ( queryHints.isEmpty() ) {
+				throw new RuntimeException( "NO HINTS!" );
```

Comment:
I believe that a comment here clarifying when this exception is thrown could improve readability.

Reply:
Fixed